src/python/pants/backend/experimental/jvm/resolve/coursier_setup.py

```diff
@@ -0,0 +1,196 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+import textwrap
+from dataclasses import dataclass
+
+from pants.core.util_rules.external_tool import (
+    DownloadedExternalTool,
+    ExternalTool,
+    ExternalToolRequest,
+)
+from pants.engine.fs import CreateDigest, Digest, FileContent, MergeDigests
+from pants.engine.platform import Platform
+from pants.engine.rules import Get, collect_rules, rule
+
+logger = logging.getLogger(__name__)
+
+
+COURSIER_POST_PROCESSING_SCRIPT = textwrap.dedent(
+    """\
+import json
+import sys
+from pathlib import PurePath
+from shutil import copyfile
+
+report = json.load(open(sys.argv[1]))
+
+classpath = set()
+for dep in report['dependencies']:
+    file_path = PurePath(dep['file'])
+    classpath_dest = f"classpath/{file_path.name}"
+    if classpath_dest in classpath:
+        raise Exception(f"Found duplicate jar name {file_path.name}, which isn't currently supported")
+    classpath.add(classpath_dest)
+    copyfile(file_path, classpath_dest)
+"""
+)
+
+
+@dataclass(frozen=True)
+class CoursierPostProcessingScript:
+    digest: Digest
+    relpath: str
+
+
+COURSIER_WRAPPER_SCRIPT = textwrap.dedent(
+    """\
+set -eux
+
+coursier_exe="$1"
+shift
+json_output_file="$1"
+shift
+
+"$coursier_exe" fetch --json-output-file="$json_output_file" "$@"
+
+/bin/mkdir -p classpath
+/usr/bin/python3 coursier_post_processing_script.py "$json_output_file"
+"""
+)
+
+JAVAC_WRAPPER_SCRIPT = textwrap.dedent(
+    """\
+set -eux
+
+coursier_exe="$1"
+shift
+
+javac_path="$(${coursier_exe} java-home --jvm openjdk:1.9.0-4)/bin/javac"
+
+exec "${javac_path}" "$@"
+"""
+)
+
+
+@dataclass(frozen=True)
+class CoursierJavac:
+    digest: Digest
+    relpath: str
+
+
+@dataclass(frozen=True)
+class CoursierWrapperScript:
+    digest: Digest
+    relpath: str
+
+
+@dataclass(frozen=True)
+class SetupCoursierRequest:
+    pass
```

Comment:
You can delete this request type. Then at callsites, rules will put `Coursier` in the function signature, rather than using `await Get`. Beyond less code, this improves concurrency because all the `@rule`'s params can be evaluated in parallel instead of blocking on a single `await Get`.

Reply:
Done